RELEASES.md

```diff
@@ -1,3 +1,193 @@
+Version 1.26.0 (2018-05-10)
+==========================
+
+Language
+--------
+- [Closures now implement `Copy + Clone` if all captured variables do.][49299]
+- [Stabilised the `!` type (A type that has no value).][47630]
```

Comment:
> A type that has no value

! isn't just a ZST (zero sized type), it has special powers. Can't recall what exactly they were but a better name would be "never type".

Reply:
Oh BTW, if we can't find a suitable fix for https://github.com/rust-lang/rust/pull/49039#issuecomment-376420816, we may be forced to unstabilize `!` again.

Reply:
I think there was a thread where it was decided that "never type" is the official, technically correct pronunciation of "`!` type", though unfortunately I can't find it. Since `!` is how you spell it in code, I'd show both names: "Stabilised the `!` type, also known as the "never" type".

Reply:
"never" is what you need to search for to find it in the docs (https://doc.rust-lang.org/nightly/std/?search=never), so it would be good to include that name here.

Reply:
> Can't recall what exactly they were but a better name would be "never type".

Because there is no value of this type, if the compiler ever sees something with this type, it must be dead code and can be optimized out.

Reply:
There are other Zero sized types like (). You can declare your own via `struct TypeName;`. This isn't the exciting thing about the ! type. It is that it is special as it *coerces to any other type*. So if you have a function `foo() -> !` you can do `let v: u32 = foo();`, or even `let v: ! = foo(); let w: u32 = v; let x: String = v;`.

Reply:
@est31 `struct TypeName;` and `()` aren't ZST's AFAIK. They're types that contain a single value. The only other way of having a ZST in Rust is having an enum with no variants. e.g. `enum Void {}`

Reply:
`struct Foo;` is definitely a ZST: https://doc.rust-lang.org/nomicon/exotic-sizes.html#zero-sized-types-zsts

Reply:
To clarify: `()` and `struct Foo;` are Zero sized types: they have exactly one possible value (`()` and `Foo`, respectively), so they can be represented with `log(1) = 0` bits, hence the name. `!` is not zero-sized; it cannot be represented with any number of bits because there are no values of that type.

Reply:
I'd say a ZST is any type T that fulfills `std::mem::size_of::<T>() == 0`. And ! does fulfill it so it is zero sized. I mean when ! has a size of zero, it must be zero-sized right?

In the "abstract" type check world, both `struct Foo;` and `enum Var{}` can have values. After all you can do things like `let v: ! = panic!(); let w: u32 = v; println!("{}", w + 1);`.

In the "technical" code gen world, both `()` and `!` get optimized out, their "values" never get onto the stack. The difference only shows if stuff is nested in stuff like `Result`, but the nomicon claims that this optimisation is not yet being performed... might be outdated given eddyb's work on it.

So I believe that the statement "a type that has no value" is neither unique to the ! type (as there are types like `enum Var {}`) nor is 100% accurate as as far as type checking is concerned it actually can.

Reply:
`And ! does fulfill it so it is zero sized.` Do you mean theoretically or in implementation? Theoretically, it should _not_ have size zero; it's size is undefined. I don't know what the implementation of `mem::size_of::<!>()` does, though. I guess 0 is probably the most reasonable answer, but really `None` or `panic!` would be more correct...

Reply:
@est31 There are *expressions* of uninhabited types (what you mean by "the type checking perspective"), but none of these expressions ever actually evaluate to a *value* (if they do via unsafe code, that's UB). So "no values" is an accurate if not unique description of `!`.

Though the specific phrasing "a type that has no value" has an unfortunate double meaning ("has no value" as in "is useless")...